In [470]:
import pandas as pd
import numpy as np

In [471]:
df_items = pd.read_csv("data/order_product.csv", sep=';')
df_items.head()

vertical_code  order_product_id                       ord_prct_src_id  \
0       LAUNDRY         468875260  d56a4990-2d0d-451b-bf99-78c710343b95   
1       LAUNDRY          60635755  0d795469-2d0e-4f6d-85ad-86d461c78c2f   
2       LAUNDRY          60722561  6d309769-449f-451f-8319-d08f57e3d839   
3       LAUNDRY         269526821  825d9206-798c-49d1-b871-a0050ea6656e   
4       LAUNDRY         185172487  601eb166-cade-4942-8b17-30ca0f3a0761   

    order_id                            ord_src_id  product_id  \
0  431926237  77fd826a-86db-4af8-ad99-1463506cb9ce    90234446   
1   73849517  06628438-f89c-4d03-830f-57f034070396    22118756   
2   74048181  0c9ab7f2-7623-4a96-978f-5f6debcb33eb    21885067   
3  258058710  0daabf93-14ae-48d1-ab2f-65329193f0fd    22192163   
4  183542172  99a35140-a333-4ee0-853a-3b0ae44d1568    22192163   

                            prct_src_id ord_prct_source_code  \
0  7b14598f-942f-4fd6-8a79-ad098809c422                order   
1  c8a735a6-cdaa-4df3-876a-326b258616ed                order   
2  1f299458-d419-4f0f-b936-92a6aba587bd                order   
3  fdc3488a-dc2d-495a-90a4-dc75ddbdbc65                order   
4  fdc3488a-dc2d-495a-90a4-dc75ddbdbc65                order   

  ord_prct_currency_code  ord_prct_quantity  ord_prct_price  \
0                    NaN                  1           35.00   
1                    NaN                  1            0.10   
2                    NaN                  1           29.20   
3                    NaN                  1            0.01   
4                    NaN                  1            0.01   

   ord_prct_price_without_vat ord_prct_creation_date_time  \
0                       35.00         2021-05-05 14:50:50   
1                        0.10         2019-08-21 19:38:26   
2                       29.20         2019-11-27 12:19:34   
3                        0.01         2020-12-15 11:55:14   
4                        0.01         2020-10-05 16:49:30   

  ord_prct_last_modified_date_time  \
0              2021-05-05 14:50:50   
1              2019-08-21 19:38:26   
2              2019-11-27 12:19:34   
3              2020-12-15 11:55:14   
4              2020-10-05 16:49:30   

                                           prct_code ord_creation_date_time  \
0  PARTNER-BR-JARDINS-JOSEMARIALISBOA-3012A122-61...    2021-05-05 14:50:50   
1  OTRO-BR-ABOLICAO-DOMHELDER-088653C0-9F44-4C01-...    2019-08-21 19:38:26   
2  OTRO-BR-CENTRO-GENERALMARCONDES-613E1618-5100-...    2019-11-27 12:19:34   
3  PARTNER-BR-SOBRADINHO-CONDOMINIOS-8253A34B-3C7...    2020-12-15 11:55:14   
4  PARTNER-BR-SOBRADINHO-CONDOMINIOS-8253A34B-3C7...    2020-10-05 16:49:30   

   order_creation_date_id ord_country_code                prtn_code  \
0                20210505               BR  JARDINS-JOSEMARIALISBOA   
1                20190821               BR       ABOLICAO-DOMHELDER   
2                20191127               BR  CENTRO-GENERALMARCONDES   
3                20201215               BR   SOBRADINHO-CONDOMINIOS   
4                20201005               BR   SOBRADINHO-CONDOMINIOS   

  ord_status_code               prct_name prct_type_code prct_type_name  \
0       DELIVERED                  Casaco        LAUNDRY     Lavandería   
1       DELIVERED  10 pc lavada e passada  UNCATEGORIZED  UNCATEGORIZED   
2       DELIVERED         Adiconal Camila          OTHER          Otros   
3       DELIVERED        COMBO 3a ENTREGA        LAUNDRY     Lavandería   
4       DELIVERED        COMBO 3a ENTREGA        LAUNDRY     Lavandería   

  prct_group_code prct_group_name  prct_price ord_channel_code  \
0         PARTNER         Partner       35.00         PROVIDER   
1            OTRO            Otro        0.10         PROVIDER   
2            OTRO            Otro       29.20         PROVIDER   
3         PARTNER         Partner        0.01         PROVIDER   
4         PARTNER         Partner        0.01         PROVIDER   

  prct_type_code_processed  partner_id      

In [472]:
df_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 201044 entries, 0 to 201043
Data columns (total 31 columns):
 #   Column                            Non-Null Count   Dtype  
---  ------                            --------------   -----  
 0   vertical_code                     201044 non-null  object 
 1   order_product_id                  201044 non-null  int64  
 2   ord_prct_src_id                   201044 non-null  object 
 3   order_id                          201044 non-null  int64  
 4   ord_src_id                        201044 non-null  object 
 5   product_id                        201044 non-null  int64  
 6   prct_src_id                       201044 non-null  object 
 7   ord_prct_source_code              201044 non-null  object 
 8   ord_prct_currency_code            6766 non-null    object 
 9   ord_prct_quantity                 201044 non-null  int64  
 10  ord_prct_price                    201044 non-null  float64
 11  ord_prct_price_without_vat        201044 non-null  f

In [473]:
df_items.prct_type_code.unique()

array(['LAUNDRY', 'UNCATEGORIZED', 'OTHER', 'IRON', 'DRYCLEANING',
       'SUBSCRIPTION'], dtype=object)

In [474]:
# Check for discrete data 

for columnae in df_orders.columns:
    print("The number of unique entries in " + str(columnae) + " is " + str(len(df_orders[columnae].unique())))
    

The number of unique entries in order_id is 122037
The number of unique entries in product_id is 54197
The number of unique entries in ord_prct_source_code is 2
The number of unique entries in ord_prct_quantity is 191
The number of unique entries in ord_prct_price is 2737
The number of unique entries in ord_prct_creation_date_time is 123359
The number of unique entries in ord_prct_last_modified_date_time is 123276
The number of unique entries in ord_creation_date_time is 121776
The number of unique entries in prtn_code is 121
The number of unique entries in ord_status_code is 10
The number of unique entries in prct_group_name is 32
The number of unique entries in ord_channel_code is 4
The number of unique entries in prct_type_code_processed is 6
The number of unique entries in partner_id is 121


Remove vertical_code, because there is only one entry for all records.

Remove ord_prct_src_id and order_product_id, because they are the same length as the dataframe, so appears to be an identifier, but ones that aren't reflected in the other dataframe. 

Remove ord_src_id, because it is the same length as the order_id, and seems to mirror the uniqueness, but order_id IS matched in the other dataframe. 

Remove prct_src_id because it is the same length as the product_id and seems to reflect the same grouping of information, if any. 

Remove ord_country_code because there is only one entry for all records. 

Remove ord_prct_currency_code, because all transactions are in the same currency. 

Remove 'prct_type_name' because it relects the same information found in prct_type_code.

Remove 'order_creation_date_id' because it is just a date code for informatin found elsewhere. 

In [475]:
# what is contained in some of the other columns?
df_items['ord_prct_source_code'].unique()

array(['order', 'provider'], dtype=object)

In [476]:
df_items['prct_type_code'].unique()

array(['LAUNDRY', 'UNCATEGORIZED', 'OTHER', 'IRON', 'DRYCLEANING',
       'SUBSCRIPTION'], dtype=object)

In [477]:
df_items['prct_type_code_processed'].unique()

array(['LAUNDRY', 'UNCATEGORIZED', 'HOME', 'IRON', 'DRYCLEANING',
       'SUBSCRIPTION'], dtype=object)

Does 'prct_type_code_processed' and 'prct_type_code' always match?

In [478]:
len(df_items.loc[df_items['prct_type_code'] != df_items['prct_type_code_processed']])


33402

It looks like one column holds "other" and the other holds "home". Let's test that. 

In [479]:
(df_items['prct_type_code'].values == 'OTHER').sum(), (df_items['prct_type_code_processed'].values == 'HOME').sum()

(32914, 32913)

Given this information, we will only keep the prct_type_code_processed and drop prct_type_code.

In [480]:
df_items.drop(['ord_prct_src_id', 'vertical_code', 'ord_country_code', 'ord_src_id',
                'order_product_id', 'prct_src_id', 'ord_prct_currency_code', 'prct_type_name', 'order_creation_date_id',
               'prtn_src_id', 'prct_type_code'],
               axis=1, inplace=True)

In [481]:
df_items.head()

order_id  product_id ord_prct_source_code  ord_prct_quantity  \
0  431926237    90234446                order                  1   
1   73849517    22118756                order                  1   
2   74048181    21885067                order                  1   
3  258058710    22192163                order                  1   
4  183542172    22192163                order                  1   

   ord_prct_price  ord_prct_price_without_vat ord_prct_creation_date_time  \
0           35.00                       35.00         2021-05-05 14:50:50   
1            0.10                        0.10         2019-08-21 19:38:26   
2           29.20                       29.20         2019-11-27 12:19:34   
3            0.01                        0.01         2020-12-15 11:55:14   
4            0.01                        0.01         2020-10-05 16:49:30   

  ord_prct_last_modified_date_time  \
0              2021-05-05 14:50:50   
1              2019-08-21 19:38:26   
2              2019-11-27 12:19:34   
3              2020-12-15 11:55:14   
4              2020-10-05 16:49:30   

                                           prct_code ord_creation_date_time  \
0  PARTNER-BR-JARDINS-JOSEMARIALISBOA-3012A122-61...    2021-05-05 14:50:50   
1  OTRO-BR-ABOLICAO-DOMHELDER-088653C0-9F44-4C01-...    2019-08-21 19:38:26   
2  OTRO-BR-CENTRO-GENERALMARCONDES-613E1618-5100-...    2019-11-27 12:19:34   
3  PARTNER-BR-SOBRADINHO-CONDOMINIOS-8253A34B-3C7...    2020-12-15 11:55:14   
4  PARTNER-BR-SOBRADINHO-CONDOMINIOS-8253A34B-3C7...    2020-10-05 16:49:30   

                 prtn_code ord_status_code               prct_name  \
0  JARDINS-JOSEMARIALISBOA       DELIVERED                  Casaco   
1       ABOLICAO-DOMHELDER       DELIVERED  10 pc lavada e passada   
2  CENTRO-GENERALMARCONDES       DELIVERED         Adiconal Camila   
3   SOBRADINHO-CONDOMINIOS       DELIVERED        COMBO 3a ENTREGA   
4   SOBRADINHO-CONDOMINIOS       DELIVERED        COMBO 3a ENTREGA   

  prct_group_code prct_group_name  prct_price ord_channel_code  \
0         PARTNER         Partner       35.00         PROVIDER   
1            OTRO            Otro        0.10         PROVIDER   
2            OTRO            Otro       29.20         PROVIDER   
3         PARTNER         Partner        0.01         PROVIDER   
4         PARTNER         Partner        0.01         PROVIDER   

  prct_type_code_processed  partner_id                    prtn_name  
0                  LAUNDRY     74752.0  Jardins - José Maria Lisboa  
1            UNCATEGORIZED     74345.0        Abolição - Dom Helder  
2                     HOME     74496.0   Centro - General Marcondes  
3                  LAUNDRY     74255.0     Sobradinho - Condominios  
4                  LAUNDRY     74255.0     Sobradinho - Condominios

In [482]:
price = df_items.loc[df_orders['ord_prct_price'] != df_items['ord_prct_price_without_vat']]
price.head(2)

order_id  product_id ord_prct_source_code  ord_prct_quantity  \
67   73661569    21963799                order                  7   
261  73582254    21963799                order                  6   

     ord_prct_price  ord_prct_price_without_vat ord_prct_creation_date_time  \
67              5.0                        4.27         2019-03-15 13:43:03   
261             5.0                        4.27         2019-03-18 12:55:44   

    ord_prct_last_modified_date_time  \
67               2019-03-15 13:43:03   
261              2019-03-18 12:55:44   

                                             prct_code ord_creation_date_time  \
67   PRENDA-SOLO-PLANCHADA-BR-PINHEIROS-CRISTIANOVIANA    2019-03-15 13:43:03   
261  PRENDA-SOLO-PLANCHADA-BR-PINHEIROS-CRISTIANOVIANA    2019-03-18 12:55:44   

                    prtn_code  ord_status_code             prct_name  \
67   PINHEIROS-CRISTIANOVIANA        DELIVERED  Roupa apenas passada   
261  PINHEIROS-CRISTIANOVIANA  ABSENT_DELIVERY  Roupa apenas passada   

           prct_group_code        prct_group_name  prct_price  \
67   PRENDA-SOLO-PLANCHADA  Prenda solo planchada         5.0   
261  PRENDA-SOLO-PLANCHADA  Prenda solo planchada         5.0   

    ord_channel_code prct_type_code_processed  partner_id  \
67           ANDROID                     IRON     74617.0   
261          ANDROID                     IRON     74617.0   

                       prtn_name  
67   Pinheiros - Cristiano Viana  
261  Pinheiros - Cristiano Viana

It looks like most of the pricing is the same for both columns, with and without VAT. Let's keep the price with VAT. and drop the column without VAT. This discrepancy affects a very small amount of the data. 

In [483]:
len(price)

773

In [484]:
df_items.drop(['ord_prct_price_without_vat'], axis=1, inplace=True)

Do 'ord_prct_creation_date_time' and 'ord_prct_last_modified_date_time' ever vary, and if so, by how much?

In [485]:
time = df_items.loc[df_orders['ord_prct_creation_date_time'] != df_items['ord_prct_last_modified_date_time']]
len(time)

739

What about the difference between ord_prct_creation_date_time and ord_creation_date_time?

In [486]:
time2 = df_items.loc[df_orders['ord_prct_creation_date_time'] != df_items['ord_creation_date_time']]
len(time2)

4264

In [487]:
time2.head()

order_id  product_id ord_prct_source_code  ord_prct_quantity  \
14   73584015    22006657             provider                  2   
82   74252211    21928280             provider                  1   
128  73968294    22006657             provider                  2   
162  83487826    21951238             provider                  1   
171  73791754    22163575             provider                  2   

     ord_prct_price ord_prct_creation_date_time  \
14             35.0         2019-10-24 13:15:38   
82             28.0         2020-03-26 22:42:51   
128            35.0         2019-12-05 15:05:57   
162            25.0         2020-07-03 15:15:44   
171             5.0         2020-03-02 17:48:24   

    ord_prct_last_modified_date_time  \
14               2019-10-24 13:15:38   
82               2020-03-26 22:42:51   
128              2019-12-05 15:05:57   
162              2020-07-03 15:15:44   
171              2020-03-02 17:48:24   

                                             prct_code ord_creation_date_time  \
14               ALFOMBRA-BR-VILAMARIANA-JOAQUIMTAVORA    2019-10-21 17:46:42   
82   PARTNER-BR-LAGONORTE-SHINCA05-F98B35AA-5FCE-45...    2020-03-19 15:38:08   
128              ALFOMBRA-BR-VILAMARIANA-JOAQUIMTAVORA    2019-12-03 15:06:16   
162                     AMERICANA-BR-NOROESTE-CLNW1011    2020-06-26 16:14:51   
171           ITAIM-BIBI-PRENDA-SOLO-PLANCHADA-BR-SUBS    2020-02-22 13:47:45   

                     prtn_code   ord_status_code             prct_name  \
14   VILAMARIANA-JOAQUIMTAVORA         DELIVERED                Tapete   
82          LAGONORTE-SHINCA05  WAITING_DELIVERY       Capa de Edredom   
128  VILAMARIANA-JOAQUIMTAVORA   ABSENT_DELIVERY                Tapete   
162          NOROESTE-CLNW1011         DELIVERED       Paletó / Blazer   
171                 ITAIM-BIBI  WAITING_DELIVERY  Roupa apenas passada   

           prct_group_code        prct_group_name  prct_price  \
14                ALFOMBRA               Alfombra        35.0   
82                 PARTNER                Partner        28.0   
128               ALFOMBRA               Alfombra        35.0   
162              AMERICANA              Americana        25.0   
171  PRENDA-SOLO-PLANCHADA  Prenda solo planchada         5.0   

    ord_channel_code prct_type_code_processed  partner_id  \
14           ANDROID                     HOME     74300.0   
82               IOS                  LAUNDRY     74678.0   
128              IOS                     HOME     74300.0   
162         PROVIDER              DRYCLEANING     75051.0   
171        ECOMMERCE             SUBSCRIPTION     74807.0   

                         prtn_name  
14   Vila Mariana - Joaquim Távora  
82         Lago Norte - Shin Ca 05  
128  Vila Mariana - Joaquim Távora  
162          Noroeste - CLNW 10 11  
171                   Itaim - Bibi

In [488]:
group = df_items.loc[df_items['prct_group_code'].str.lower() != df_items['prct_group_name'].str.lower()]

In [489]:
len(group)

55500

In [490]:
group.head()

order_id  product_id ord_prct_source_code  ord_prct_quantity  \
16   73957351    21993789                order                  1   
18  221984438    21876881                order                  1   
24  141252137    21876881                order                  1   
27   73544984    21876881                order                  2   
28  242550337    78351579                order                  1   

    ord_prct_price ord_prct_creation_date_time  \
16            42.0         2020-06-08 17:04:15   
18            25.0         2020-11-11 17:58:25   
24            25.0         2020-08-28 12:01:07   
27            25.0         2019-10-15 17:19:31   
28            65.0         2020-12-01 23:44:24   

   ord_prct_last_modified_date_time  \
16              2020-06-08 17:04:15   
18              2020-11-11 17:58:25   
24              2020-08-28 12:01:07   
27              2019-10-15 17:19:31   
28              2020-12-01 23:44:24   

                                     prct_code ord_creation_date_time  \
16         EDREDON-QUEEN-SIZE-BR-CANOAS-CENTRO    2020-06-08 17:04:15   
18      EDREDON-FULL-SIZE-BR-NOROESTE-CLNW1011    2020-11-11 17:58:25   
24      EDREDON-FULL-SIZE-BR-NOROESTE-CLNW1011    2020-08-28 12:01:07   
27      EDREDON-FULL-SIZE-BR-NOROESTE-CLNW1011    2019-10-15 17:19:31   
28  BOLSA-LAVANDERIA-BR-COPACABANA-FRANCISCOSA    2020-12-01 23:44:24   

                 prtn_code ord_status_code                 prct_name  \
16           CANOAS-CENTRO       DELIVERED  Edredom Casal Queen Size   
18       NOROESTE-CLNW1011       DELIVERED             Edredom Casal   
24       NOROESTE-CLNW1011       DELIVERED             Edredom Casal   
27       NOROESTE-CLNW1011       DELIVERED             Edredom Casal   
28  COPACABANA-FRANCISCOSA       DELIVERED       Bolsa de Lavanderia   

       prct_group_code     prct_group_name  prct_price ord_channel_code  \
16  EDREDON-QUEEN-SIZE  Edredón queen size        42.0         PROVIDER   
18   EDREDON-FULL-SIZE   Edredón full size        25.0         PROVIDER   
24   EDREDON-FULL-SIZE   Edredón full size        25.0         PROVIDER   
27   EDREDON-FULL-SIZE   Edredón full size        25.0              IOS   
28    BOLSA-LAVANDERIA    Bolsa lavanderia        65.0              IOS   

   prct_type_code_processed  partner_id                  prtn_name  
16                     HOME     74142.0            Canoas - Centro  
18                     HOME     75051.0      Noroeste - CLNW 10 11  
24                     HOME     75051.0      Noroeste - CLNW 10 11  
27                     HOME     75051.0      Noroeste - CLNW 10 11  
28                  LAUNDRY    110713.0  Copacabana - Francisco Sá

We can see that these categories are the same except for hyphens and accents. 

In [491]:
# drop similar columns
df_items.drop(['prct_group_code', 'prct_name'], axis=1, inplace=True)

In [492]:
pricing = df_items.loc[df_items['ord_prct_price'] != df_items['prct_price']]
len(pricing)

1005

In [493]:
# drop prct_price
df_items = df_items.drop(['prct_price'], axis=1)

In [494]:
# drop prct_code because it duplicates other information but in a more complicated way
df_items = df_items.drop(['prct_code', 'prtn_name'], axis=1)

In [495]:
df_items.head()

order_id  product_id ord_prct_source_code  ord_prct_quantity  \
0  431926237    90234446                order                  1   
1   73849517    22118756                order                  1   
2   74048181    21885067                order                  1   
3  258058710    22192163                order                  1   
4  183542172    22192163                order                  1   

   ord_prct_price ord_prct_creation_date_time  \
0           35.00         2021-05-05 14:50:50   
1            0.10         2019-08-21 19:38:26   
2           29.20         2019-11-27 12:19:34   
3            0.01         2020-12-15 11:55:14   
4            0.01         2020-10-05 16:49:30   

  ord_prct_last_modified_date_time ord_creation_date_time  \
0              2021-05-05 14:50:50    2021-05-05 14:50:50   
1              2019-08-21 19:38:26    2019-08-21 19:38:26   
2              2019-11-27 12:19:34    2019-11-27 12:19:34   
3              2020-12-15 11:55:14    2020-12-15 11:55:14   
4              2020-10-05 16:49:30    2020-10-05 16:49:30   

                 prtn_code ord_status_code prct_group_name ord_channel_code  \
0  JARDINS-JOSEMARIALISBOA       DELIVERED         Partner         PROVIDER   
1       ABOLICAO-DOMHELDER       DELIVERED            Otro         PROVIDER   
2  CENTRO-GENERALMARCONDES       DELIVERED            Otro         PROVIDER   
3   SOBRADINHO-CONDOMINIOS       DELIVERED         Partner         PROVIDER   
4   SOBRADINHO-CONDOMINIOS       DELIVERED         Partner         PROVIDER   

  prct_type_code_processed  partner_id  
0                  LAUNDRY     74752.0  
1            UNCATEGORIZED     74345.0  
2                     HOME     74496.0  
3                  LAUNDRY     74255.0  
4                  LAUNDRY     74255.0

In [496]:
df_items.prct_group_name.unique()

array(['Partner', 'Otro', 'Alfombra', 'Edredón queen size',
       'Edredón full size', 'Americana', 'Manta', 'Bolsa lavanderia',
       'Prenda lavada y planchada', 'Edredón king size',
       'Prenda solo planchada', 'Vestido de fiesta',
       'Tenis de cuero o gamuza', 'Jersey', 'Media bolsa lavandería',
       'Decena de prendas lavadas y planchadas', 'Juego de cama',
       'Pantalon', 'Decena de prendas planchadas', 'Traje 2 piezas',
       'Cortina', 'Tenis', 'Corbata', 'Camisa', 'Abrigo',
       'Edredon sintetico', 'Prenda planchada', 'Edredón de single size',
       'Juego de toallas', 'Prenda solo lavada', 'Camisa planchada',
       'Camisa Lavada-Planchada'], dtype=object)

Load the other dataframe and join. 
This join will be a left join on the df_orders, on the order_id column. 


In [497]:
# Are order IDs unique in this dataset? 
len(df_items), len(df_items['order_id'].unique())

(201044, 122037)

In [498]:
# Are product IDs unique in this dataset? 
len(df_items), len(df_items['product_id'].unique())

(201044, 54197)

It looks like the only thing that is unique to each record in this dataframe is the indexer. 

Order id's are not unique. Several records may be associated with the same order_id. We want the dataframe to be joined on *each* of those instances. Not just one of the instances. 

In [499]:
# load cleaned df
df = pd.read_csv('data/orderskpi.csv')
df.info()

# Note in this dataset, order_id is unique. 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149356 entries, 0 to 149355
Data columns (total 16 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   partner_id               149356 non-null  int64  
 1   customer_id              149356 non-null  int64  
 2   ord_type_code            149356 non-null  object 
 3   order_id                 149356 non-null  int64  
 4   order_date               149356 non-null  object 
 5   first_purchase           149356 non-null  bool   
 6   number_of_purchases      149356 non-null  int64  
 7   ord_channel_type_code    149356 non-null  object 
 8   ord_payment_status_code  149356 non-null  object 
 9   sales                    149356 non-null  float64
 10  pymt_order               149356 non-null  float64
 11  refund_order             149356 non-null  float64
 12  couponUsed               149356 non-null  int64  
 13  dayIndex                 149356 non-null  int64  
 14  isWe

We need a many-to-one join. In df_items, there are unique values versus the total rows:

In [500]:
len(df_items['order_id'].unique()), len(df_items)

(122037, 201044)

In df, there is a unique order_id for each row:

In [501]:
len(df['order_id'].unique().tolist()), len(df)

(149356, 149356)

How many of those order_ids overlap?

In [502]:
df1= pd.DataFrame((df_items.order_id.isin(df.order_id)))

In [503]:
df1[df1.order_id == True].count()

order_id    174973
dtype: int64

In [504]:
x = df_items['order_id'].unique().tolist()
y = df['order_id'].unique().tolist()

l_func = lambda x, y: list((set(x)- set(y))) + list((set(y)- set(x))) 


non_match = l_func(x, y)

In [505]:
print(174973 - len(non_match))

123912


The 'many' is in df_items, and the one is in df. We should get a dataframe of length 174973 as shown above. 

In [506]:
merged = pd.merge(df_items, df, validate ="many_to_one")

In [507]:
# settings to display all columns
pd.set_option("display.max_columns", None)

In [508]:
merged.head(10)

order_id  product_id ord_prct_source_code  ord_prct_quantity  \
0  431926237    90234446                order                  1   
1   73849517    22118756                order                  1   
2   74048181    21885067                order                  1   
3  258058710    22192163                order                  1   
4  183542172    22192163                order                  1   
5  290373874    21857882                order                  5   
6  290373874    22050092                order                  1   
7  290373874    22179732                order                  1   
8  290373874    22058133                order                  4   
9  338429262   127154118                order                  1   

   ord_prct_price ord_prct_creation_date_time  \
0           35.00         2021-05-05 14:50:50   
1            0.10         2019-08-21 19:38:26   
2           29.20         2019-11-27 12:19:34   
3            0.01         2020-12-15 11:55:14   
4            0.01         2020-10-05 16:49:30   
5            7.30         2021-01-11 18:03:30   
6            8.00         2021-01-11 18:03:30   
7            4.50         2021-01-11 18:03:30   
8            4.50         2021-01-11 18:03:30   
9           10.00         2021-02-22 17:52:18   

  ord_prct_last_modified_date_time ord_creation_date_time  \
0              2021-05-05 14:50:50    2021-05-05 14:50:50   
1              2019-08-21 19:38:26    2019-08-21 19:38:26   
2              2019-11-27 12:19:34    2019-11-27 12:19:34   
3              2020-12-15 11:55:14    2020-12-15 11:55:14   
4              2020-10-05 16:49:30    2020-10-05 16:49:30   
5              2021-01-11 18:03:30    2021-01-11 18:03:30   
6              2021-01-11 18:03:30    2021-01-11 18:03:30   
7              2021-01-11 18:03:30    2021-01-11 18:03:30   
8              2021-01-11 18:03:30    2021-01-11 18:03:30   
9              2021-02-22 17:52:18    2021-02-22 17:52:18   

                 prtn_code   ord_status_code prct_group_name ord_channel_code  \
0  JARDINS-JOSEMARIALISBOA         DELIVERED         Partner         PROVIDER   
1       ABOLICAO-DOMHELDER         DELIVERED            Otro         PROVIDER   
2  CENTRO-GENERALMARCONDES         DELIVERED            Otro         PROVIDER   
3   SOBRADINHO-CONDOMINIOS         DELIVERED         Partner         PROVIDER   
4   SOBRADINHO-CONDOMINIOS         DELIVERED         Partner         PROVIDER   
5                MACAE-SUL  WAITING_DELIVERY         Partner         PROVIDER   
6                MACAE-SUL  WAITING_DELIVERY         Partner         PROVIDER   
7                MACAE-SUL  WAITING_DELIVERY         Partner         PROVIDER   
8                MACAE-SUL  WAITING_DELIVERY         Partner         PROVIDER   
9                MACAE-SUL    WAITING_PICKUP            Otro         PROVIDER   

  prct_type_code_processed  partner_id  customer_id ord_type_code  order_date  \
0                  LAUNDRY     74752.0    290277393           B2C  2021-05-05   
1            UNCATEGORIZED     74345.0     75077347           B2C  2019-08-21   
2                     HOME     74496.0     75446070           B2C  2019-11-27   
3                  LAUNDRY     74255.0     74900061           B2C  2020-12-15   
4                  LAUNDRY     74255.0     74913075           B2C  2020-10-05   
5                  LAUNDRY     74157.0     75113373           B2C  2021-01-11   
6                  LAUNDRY     74157.0     75113373           B2C  2021-01-11   
7                  LAUNDRY     74157.0     75113373           B2C  2021-01-11   
8                  LAUNDRY     74157.0     75113373           B2C  2021-01-11   
9                  LAUNDRY     74157.0    536345477           B2C  2021-02-22   

   first_purchase  number_of_purchases ord_channel_type_code  \
0           False                    6               OFFLINE   
1           False                    9               OFFLINE   
2           False                    5               OFFLINE   
3  

In [509]:
len(merged)

174973

Our database is the correct length - the merge was performed correctly.
After consult with client representative, drop all dates except ddate. 

In [510]:
df_ml = merged.drop(['ord_prct_creation_date_time', 
                      'ord_creation_date_time', 'ord_prct_last_modified_date_time'], axis=1)

Some subscription orders remain in the list. Remove them:

In [511]:
# remove records within 'prct_type_code_processed' which are on a subscription customer
df_ml = df_ml[df_ml['prct_type_code_processed'] != 'SUBSCRIPTION']

In [512]:
# remove order type Subscription
df_ml = df_ml[df_ml['ord_type_code'] != 'SUBSCRIPTION']

In [513]:
# confirm no other data resides in 'ord_type_code'
df_ml['ord_type_code'].unique()

array(['B2C'], dtype=object)

In [514]:
# drop ord_type_code
df_ml.drop(columns=['ord_type_code'], axis=1, inplace=True)

In [515]:
# Write the cleaned df to file. 
df_ml.to_csv('data/merged_df.csv', index=False)

In [516]:
df_ml = df_ml.sort_values(by='order_id')

In [517]:
df_ml.columns

Index(['order_id', 'product_id', 'ord_prct_source_code', 'ord_prct_quantity',
       'ord_prct_price', 'prtn_code', 'ord_status_code', 'prct_group_name',
       'ord_channel_code', 'prct_type_code_processed', 'partner_id',
       'customer_id', 'order_date', 'first_purchase', 'number_of_purchases',
       'ord_channel_type_code', 'ord_payment_status_code', 'sales',
       'pymt_order', 'refund_order', 'couponUsed', 'dayIndex', 'isWeekday',
       'serviceChannel'],
      dtype='object')

In [518]:
len(df_ml.order_id.unique())

109005

In [519]:
df_ml.head(20)

order_id  product_id ord_prct_source_code  ord_prct_quantity  \
159916  73542482    22110820                order                  1   
59732   73542488    22190247                order                  1   
132220  73542515    22046197                order                  1   
16320   73542618    21931132                order                  1   
16319   73542618    22184402                order                  1   
151744  73542619    21994420                order                  1   
143873  73542693    21917080                order                  1   
69280   73542694    21854446                order                  1   
153951  73542713    22028187                order                  1   
90570   73542818    21978174                order                  1   
131643  73542955    21872301                order                  6   
157389  73542956    21980038                order                  1   
115355  73543001    21885105                order                  1   
98996   73543002    22028833                order                  2   
104119  73543063    22166102                order                  1   
96303   73543064    22129923                order                  3   
96304   73543064    22055338                order                  4   
96305   73543064    22079071                order                  1   
139620  73543111    21922933                order                  2   
139621  73543111    21890931                order                  1   

        ord_prct_price                         prtn_code   ord_status_code  \
159916            60.0         JARDIMSUMARE-JOAOPANTEADO         DELIVERED   
59732             39.0                 ZONANORTE-AFLITOS         DELIVERED   
132220            39.0          FREGUESIA-LINHAAMARELA1B  WAITING_DELIVERY   
16320             39.0                LAGONORTE-SHINCA05         DELIVERED   
16319             60.0                LAGONORTE-SHINCA05         DELIVERED   
151744            60.0       BOAVIAGEM-CONSELHEIROAGUIAR         DELIVERED   
143873            39.0  CHACARAPRIMAVERA-JORGEFIGUEIREDO         DELIVERED   
69280             38.0              VILACLEMENTINO-ROSAS         DELIVERED   
153951            39.0          CONTINENTE-SANTOSSARAIVA           WASHING   
90570             35.0               BATEL-COSTACARVALHO         DELIVERED   
131643             8.0            SAOJUDAS-FAGUNDESFILHO         DELIVERED   
157389            35.0                     SUDOESTE-300A         DELIVERED   
115355            39.0            ALTOXV-MARECHALDEODORO         DELIVERED   
98996             60.0                  BAURU-VILAGUEDES         DELIVERED   
104119            38.0                 JARDIMEUROPA-VIVA         DELIVERED   
96303             38.0                 BELAVISTA-ALAMEDA         DELIVERED   
96304              6.0                 BELAVISTA-ALAMEDA         DELIVERED   
96305            110.0                 BELAVISTA-ALAMEDA         DELIVERED   
139620            35.0               BATEL-COSTACARVALHO         DELIVERED   
139621            30.0               BATEL-COSTACARVALHO         DELIVERED   

               prct_group_name ord_channel_code prct_type_code_processed  \
159916        Bolsa lavanderia         PROVIDER                  LAUNDRY   
59732   Media bolsa lavandería              IOS                  LAUNDRY   
132220  Media bolsa lavandería          ANDROID                  LAUNDRY   
16320   Media bolsa lavandería          ANDROID                  LAUNDRY   
16319         Bolsa lavanderia          ANDROID                  LAUNDRY   
151744        Bolsa lavanderia          ANDROID                  LAUNDRY   
143873  Media bolsa lavandería          ANDROID                  LAUNDRY   
69280        Edredón full size          ANDROID                     HOME   
153951  Media bolsa lavandería         PROVIDER                  LAUNDRY   
90570       Edredón queen size              IOS                     HOME   
131643   Prenda solo

For each order_id, much of this should be the same. Each order_id is attached to multiple products. Each product *should* have the same number of purchases, channel code, etc. We would like to combine all into one group so that each order_id is associated with several items. 

Those categories are:

    1. Group the prct_group_name entries into a list for each order_id. 
    
    2. Add together entries in the ord_prct_quantity column to sum total items by order_id. 
    
    3. Are there any orders which are partially fulfilled? Use agg func 'nunique' to count number of unique entries within 'ord_status_code' by 'order_id'. 
    
    4. Are there any orders which are partially paid? Use agg func 'nunique' to count number of unique entries within 'ord_payment_status_code' by 'order_id'. 
    
    5. Are there orders which have different processing codes? First use agg func 'nunique' to count number of unique entries within 'prct_type_code_processed'. If there are more than one, use agg func 'unique' to aggregate unique entries by 'order_id'. 
    
    6. Make a small table of the price for each item - which is cheapest to most expensive. 

1. Group the prct_group_name entries into a list for each order_id. 

In [520]:
# group the items ordered by order_id. 
stuff_ordered = df_ml.groupby('order_id')['prct_group_name'].agg('unique')

# check that the length is equal to the dataframe's number of unique order ids. 
len(stuff_ordered.index.unique())

109005

In [521]:
# show the dataframe
stuff_ordered.head()

order_id
73542482                            [Bolsa lavanderia]
73542488                      [Media bolsa lavandería]
73542515                      [Media bolsa lavandería]
73542618    [Media bolsa lavandería, Bolsa lavanderia]
73542619                            [Bolsa lavanderia]
Name: prct_group_name, dtype: object

2. Add together entries in the ord_prct_quantity column to sum items by order_id.

In [522]:
# Add up the number of items purchased
NumOfItems = df_ml.groupby('order_id')['ord_prct_quantity'].agg('sum')  
NumOfItems.head()

order_id
73542482    1
73542488    1
73542515    1
73542618    2
73542619    1
Name: ord_prct_quantity, dtype: int64

3. Are there any orders which are partially fulfilled? Use agg func 'nunique' to count number of unique entries within 'ord_status_code' by 'order_id'. 

In [523]:
# group the fulfillment of orders by order_id. 
fulfilled = df_ml.groupby('order_id')['ord_status_code'].agg('nunique')
fulfilled.unique()

array([1], dtype=int64)

There are not any orders with multiple entries for the ord_status_code. 

4. ord_payment_status_code - are there any orders which are partially paid? Use agg func 'unique to count the number of unique entries by 'order_id'. 

In [524]:
# group the payment status of orders by order_id. 
paid = df_ml.groupby('order_id')['ord_payment_status_code'].agg('nunique')
paid.unique()

array([1], dtype=int64)

There are not any orders with multiple entries for the ord_status_code. 

5. Are there orders which have different processing codes? First use agg func 'nunique' to count number of unique entries within 'prct_type_code_processed'. If there are more than one, use agg func 'unique' to aggregate unique entries by 'order_id'.

In [525]:
# group the processing codes of orders by order_id. 
processed = df_ml.groupby('order_id')['prct_type_code_processed'].agg('nunique')
processed.unique()

array([1, 2, 3, 4, 5], dtype=int64)

In [526]:
processed = df_ml.groupby('order_id')['prct_type_code_processed'].agg('unique')

In [527]:
processed.head()

order_id
73542482    [LAUNDRY]
73542488    [LAUNDRY]
73542515    [LAUNDRY]
73542618    [LAUNDRY]
73542619    [LAUNDRY]
Name: prct_type_code_processed, dtype: object

There are orders which have different processing codes. These codes are now aggregated in a list for each order_id. 

6. Make a small table of the price for each item - which is cheapest to most expensive. 

In [528]:
price = df_ml[['ord_prct_price', 'prct_group_name']]
price['prct_group_name'].unique()

array(['Bolsa lavanderia', 'Media bolsa lavandería', 'Edredón full size',
       'Edredón queen size', 'Prenda solo planchada', 'Traje 2 piezas',
       'Manta', 'Partner', 'Alfombra', 'Corbata',
       'Prenda lavada y planchada', 'Otro', 'Camisa',
       'Decena de prendas lavadas y planchadas', 'Edredón de single size',
       'Decena de prendas planchadas', 'Edredon sintetico', 'Americana',
       'Vestido de fiesta', 'Pantalon', 'Juego de cama',
       'Tenis de cuero o gamuza', 'Abrigo', 'Cortina', 'Tenis',
       'Edredón king size', 'Jersey', 'Prenda planchada',
       'Juego de toallas'], dtype=object)

In [529]:
items_dict = {'Bolsa lavanderia':'Laundry bag', 'Media bolsa lavanderia':'Half laundry bag', 'Edredón full size':'Quilt full size',
             'Edredón queen size':'Quilt queen size', 'Prenda solo planchada':'Garment only ironed', 'Traje 2 piezas':'2-piece suit',
              'Partner':'Partner', 'Manta':'Blanket', 'Alfombra':'Carpet', 'Corbata':'Necktie','Prenda lavada y planchada':'Garment washed and ironed',
              'Otro':'Other', 'Camisa':'Shirt', 'Decena de prendas lavadas y planchadas':'Ten items washed and ironed',
              'Edredón de single size': 'Quilt single size','Decena de prendas planchadas':'Ten ironed garments',
              'Edredon sintetico':'Quilt synthetic',  'Americana':'American','Vestido de fiesta':'Party dress', 'Pantalon':'Trousers',
              'Juego de cama':'Bedding set','Tenis de cuero o gamuza':'Leather or suede sneakers', 'Abrigo':'Coat',
              'Cortina':'Curtain', 'Tenis':'Sneakers','Edredón king size':'Quilt king size', 'Jersey':'Sweater',
              'Prenda planchada':'Ironed garment','Juego de toallas':'Set of towels', 'Prenda solo lavada':'Garment only washed',
              'Camisa Lavada-Planchada':'Washed-ironed shirt', 'Camisa planchada':'Ironed shirt'}

In [530]:
items = pd.DataFrame(list(items_dict.items()), columns=['Spanish','English']) 
len(items)

32

In [531]:
# get prices using mean
prices = round(df_ml.groupby('prct_group_name')['ord_prct_price'].agg('mean'), 0)

In [532]:
prices = prices.sort_values()
prices

prct_group_name
Prenda solo planchada                      7.0
Corbata                                   10.0
Prenda lavada y planchada                 11.0
Prenda planchada                          11.0
Camisa                                    12.0
Pantalon                                  13.0
Jersey                                    16.0
Cortina                                   20.0
Partner                                   25.0
Americana                                 25.0
Juego de toallas                          25.0
Juego de cama                             28.0
Edredón de single size                    28.0
Tenis                                     30.0
Manta                                     32.0
Edredón full size                         33.0
Alfombra                                  34.0
Edredon sintetico                         35.0
Edredón queen size                        36.0
Tenis de cuero o gamuza                   39.0
Traje 2 piezas                            39

Combine several of these dataframes:

stuff_ordered, NumOfItems and processed


In [533]:
#agg = stuff_ordered.merge(NumOfItems, on = ['order_id'],how = 'outer').merge(processed,  on = ['order_id'], how = 'outer')

agg = pd.concat([stuff_ordered, NumOfItems, processed], axis=1)
agg.reset_index(inplace=True)
len(agg)

109005

Now that we have this aggregation of the multiple items from each order, we need to join it to the rest of the df_ml dataframe. 

    'ord_prct_source_code' - This is either 'order' or 'provider'

    'ord_status_code' - This is one of the following: 'DELIVERED', 'WAITING_DELIVERY', 'WASHING', 'WASHED', 'PICKED_UP',
       'WAITING_PICKUP', 'ABSENT_DELIVERY', 'ABSENT_PICKUP', 'CREATED','CANCELLED'
       
    'ord_channel_code'- This is one of the following: 'PROVIDER', 'IOS', 'ANDROID', 'ECOMMERCE'
       
    'partner_id' - This is a partner_id, there are 82 different partners

    'customer_id' - This is customer_id, there are 34,112 different customers

    'order_date' - This is the order date

    'first_purchase' - This is True or False

    'number_of_purchases' - This ranges from 1 to 1302

    'ord_channel_type_code' - This is Offline or Online

    'ord_payment_status_code' - This is one of the following: 'PAID', 'UNPAID', 'UNPAID_PARTIAL'

    'pymt_order' - Payment

    'refund_order' - Refund amount, if any

    'couponUsed' - 1 for coupon used, otherwise 0

    'isWeekday'- 1 if the order was on a weekday


In [534]:
columns_to_keep = ['order_id', 'customer_id', 'order_date', 'partner_id', 'ord_prct_source_code', 'ord_status_code',
          'ord_channel_code', 'ord_channel_type_code','pymt_order', 'refund_order', 'couponUsed',
          'first_purchase', 'number_of_purchases', 'isWeekday']


In [535]:
agg.columns

Index(['order_id', 'prct_group_name', 'ord_prct_quantity',
       'prct_type_code_processed'],
      dtype='object')

There are 109005 unique order_ids in both the df_ml and the agg dataframe. When we join, there should be that many rows. Check that they are the same values:

In [536]:
# are the arrays comprising the unique order_ids the same between df_ml and agg
np.unique(df_ml.order_id.unique() == agg.order_id.unique())

array([ True])

In [537]:
# filter the columns we wish to keep and drop duplicated of 'order_id'
to_merge = df_ml[columns_to_keep]
to_merge = to_merge.drop_duplicates(subset=['order_id'], keep='first')

In [538]:
final = pd.merge(to_merge, agg, on='order_id', how='left')
len(final)

109005

In [539]:
# Write the cleaned df to file. 
final.to_csv('data/agg_df.csv', index=False)

In [540]:
# write the prices df t0 file
prices.to_csv('data/prices.csv', index=False)

In [541]:
final.head()

order_id  customer_id  order_date  partner_id ord_prct_source_code  \
0  73542482     75151090  2019-11-12     75065.0                order   
1  73542488     74700288  2019-08-27     74402.0                order   
2  73542515     75550863  2019-07-25     74612.0                order   
3  73542618     75260852  2020-01-08     74678.0                order   
4  73542619     74711088  2019-12-23     74084.0                order   

    ord_status_code ord_channel_code ord_channel_type_code  pymt_order  \
0         DELIVERED         PROVIDER               OFFLINE         0.0   
1         DELIVERED              IOS                ONLINE         9.0   
2  WAITING_DELIVERY          ANDROID                ONLINE         0.0   
3         DELIVERED          ANDROID                ONLINE        49.5   
4         DELIVERED          ANDROID                ONLINE        34.0   

   refund_order  couponUsed  first_purchase  number_of_purchases  isWeekday  \
0           0.0           1            True                    1          1   
1           0.0           1           False                    2          1   
2           0.0           1            True                    1          1   
3           0.0           1           False                   12          1   
4           0.0           1           False                    2          1   

                              prct_group_name  ord_prct_quantity  \
0                          [Bolsa lavanderia]                  1   
1                    [Media bolsa lavandería]                  1   
2                    [Media bolsa lavandería]                  1   
3  [Media bolsa lavandería, Bolsa lavanderia]                  2   
4                          [Bolsa lavanderia]                  1   

  prct_type_code_processed  
0                [LAUNDRY]  
1                [LAUNDRY]  
2                [LAUNDRY]  
3                [LAUNDRY]  
4                [LAUNDRY]